system/include/emscripten/proxying.h

```diff
@@ -88,6 +88,20 @@ int emscripten_proxy_callback(em_proxying_queue* q,
                               void (*cancel)(void*),
                               void* arg);
 
+// Enqueue `func` on the given queue and thread. Once (and if) it finishes the
+// task by calling `emscripten_proxy_finish` on the given `em_proxying_ctx`, it
+// will asynchronously proxy `callback` back to the current thread on the same
+// queue, or if the target thread dies before the work can be completed,
+// `cancel` will be proxied back instead. All three function will receive the
+// same argument, `arg`. Returns 1 if `func` was successfully enqueued and the
+// target thread notified or 0 otherwise.
+int emscripten_proxy_callback_with_ctx(em_proxying_queue* q,
```

Comment:
Should we just called this  `emscripten_proxy_async_with_ctx`?

I think maybe the `_with_callback` distinction is important, and also obvious the from the args.    Ideally I think we could remove the `_with_callback` variants that just have the normal versions take optional callbacks in all cases.

Another reason is that all the current function start with either `emscripten_proxy_sync` or `emscripten_proxy_async`

Reply:
After this full sequence of PRs, we have :

```
emscripten_proxy_async
emscripten_proxy_sync{_with_ctx}
emscripten_proxy_callback{_with_ctx}
emscripten_proxy_promise{_with_ctx}
```

If we really wanted to minimize the API surface, my preference would be to remove `emscripten_proxy_async` and `emscripten_proxy_callback*`, leaving only `sync` and `promise` proxying, with and without `ctx`s. The reason I haven't tried to do that minimization yet is that it seems low cost to expose the extra functions if we're going to have nearly identical functions under the hood either way. Users get the most options at negligible maintenance cost to us.

Reply:
Sounds reasonable.

As a general rule, I would rather keep the low level C API as small as possible.  For this kind of low level API (which I hope most users will never need to use), I think its fine if we don't include helpers/wrappers/shortcuts.

If we want to provide them in C++ API, or in a higher level set of C APIS, that would be fine, but at least for me it helper conceptually to keep this surface small... but maybe thats just me.

None of this has to block this change..  I don't think refactoring this layer later on is very risky.